### ISSUE TITLE
Include optional support for Dispatcher no-cache header for ACS Commons Component Error Handler

### ISSUE THREAD

**Issue by markus-meisterernst** [Adobe-Consulting-Services/acs-aem-commons#972]
Hi,

We realized the usefulness of the Component Error Handler to support a gentle degeneration of a page in the context of a partial failure in one or more AEM/Sling Components !

https://adobe-consulting-services.github.io/acs-aem-commons/features/component-error-handler/index.html

See the [forum discussion for more context information](http://help-forums.adobe.com/content/adobeforums/en/experience-manager-forum/adobe-experience-manager.topic.html/forum__axlp-hi_im_neithera.html).

One Feature that is lacking, is support for excluding the degenerated page from being cached by the AEM Dispatcher. Something like:

```
private void preventDispatcherCache(HttpServletResponse httpResponse) {
      if(!httpResponse.isCommitted()) {
         httpResponse.setHeader("Dispatcher","no-cache");
      } else {
         LOG.warn("could not set 'Dispatcher: no-cache' header, because response was already commited.");
      }
}
```

being called from ```handleError(...)``` private method, if some flag, let's say ```preventCachingOnDegeneration``` is set to true.

By means of the flag it would be configurable for the Customer use case at hand.

Thanks !

Cheers,

Markus

**Comment by bstopp** [comment]
The pages should be responding with a non-200 response code, which isn't cached by the dispatcher. Is this not the case?

**Comment by justinedelson** [comment]
@markus-meisterernst does this work in practice? I would think that in most non-trivial cases, the `setHeader()` call you have there would happen after the response body has started to be sent and that code would generate an exception since you can't set headers after the body has started to be sent.

**Comment by davidjgonzalez** [comment]
@justinedelson the isCommitted() check should prevent exceptions, no? (maybe that was adde post-comment?) IMO this is ~ the same problem we have w/ error page handler. If the error is far enough down the page, the response is committed and we cant do much wrt to error handling (since the status is already flushed).

@markus-meisterernst have you seen alot of use-cases where a component intermittently fails? IME, unless its something that is taking into account auth-status, the component is going to keep failing until either the content or code changes (both of which should flush the cache).

@bstopp this is component error handler, not page error handler, so we've always responded with 200 for these.

**Comment by justinedelson** [comment]
@davidjgonzalez ah, I missed that part in the patch... but my question still stands. In practice, would the header ever get set by that block? Or would the response almost always be committed by that point?

**Comment by markus-meisterernst** [comment]
@davidjgonzalez it happens once in a while when working with instable backends. I have to admit, that it's a corner case, as usually everthing should be stable and Page initial markup with NO personal or highly dynamic content should have already being cached. 
But, when the cached content has vanished (one of the content authors publish changes, that happens basically multiple times every day) it is even more important to not have an All or Nothing situation when the backend is unstable (I guess you know Murpy ...), so to not fail completely.

@justinedelson AFAIK is the response committed upon calling flush() on the Writer or when the underlying Buffer is filled up (the buffering and the default sizes of buffers are to some extend Servlet Container specific).
I checked with the Servlet 2.4 Spec: http://www.dmi.usherb.ca/~frappier/ift287/servlet-2_4-fr-spec.pdf (pages 43-45).
So the answer is as "always" - it depends.

But I'm not aware of the nifty details of how Sling fuses together all the Script Output that makes up a page. I guess it uses some Response Wrappers internally.

As for our CQ 5.6.1 based JSP Application that uses the suggested Code to prevent Caching in the context of pratial failure, I cannot remember to have seen such an "could not set 'Dispatcher: no-cache' header, because response was already commited."  in the logfile.
But we have no efficient Log File analysis for such cases, so it is not a clear evidence.

**Comment by benpeter** [comment]
I'd argue that if a positive result is cacheable (i.e. doesn't depend on point of time, session / user), data shouldn't be pulled from a backend at render time in the first place (rather pulled from the backend into the repo on a separate schedule and managing backend instability there), but may be missing the point.

**Comment by benpeter** [comment]
also that view of course doesn't remove the need you describe if backend at render time is what you have to deal with

**Comment by markus-meisterernst** [comment]
@benpeter yes and no:

yes, it boils down to the matter of how the architecture of a system is layed out in the first place or during the construction of the site.
 
No, As the project was ramped up with tons of consultants with their own specific view on CQ and a "Feature First !" strategy. They were on a thight schedule, so some of the cross-cutting conerns got neglected to some extend. 
In Addition to this, some requirements were revealed quite late in the implementation process.
Basically the team of today is not the team when the system was built.
All this are common "Anti"-Patterns that appear in quite a view projects I guess... unfortunately.

So, from my perspective, the enhancement I requested here would be for those of the rest of us that do not have the perfect world in place, however, try to push the limits further by implying technology where justifiable (in this case the ACS Commons library).
